### PR TITLE
ci: push rook image to other oci like quay,ghcr

### LINKED ISSUE #14833: Mirror release images to other repos such as quay.io and ghcr.io
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
To allow for flexibility in the source of downloading the Rook images, we should consider mirroring the Rook images to repos in addition to dockerhub, for example, quay.io and ghcr.io.

**What is use case behind this feature:**
This will allow users to consume images from sources that are compliant with company policies that may be locking down some sources.

### PR DESCRIPTION
**Pull request by subhamkrai**
currently rook images are only available in docker, let's start pushing rook images to quay.io and ghcr.io as well.

**Issue resolved by this Pull Request:**
Resolves #14833

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by subhamkrai** [comment]
moving it to draft as we still need to make changes according to the requirements.

File: tests/scripts/test-ci.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    amd64_Release_Image=$(docker images --format "{{.Repository}}:{{.Tag}}" | grep "amd64" | grep -v "amd64:latest" | grep -v "amd64:master")
+    tag_amd64_Release_Image="$(echo "$amd64_Release_Image" | cut -d ":" -f 2)"
+    docker tag "$amd64_Release_Image" quay.io/rook/rook-amd64:"$tag_amd64_Release_Image"
+    docker push quay.io/rook/rook-amd64:"$tag_amd64_Release_Image"
+    docker tag "$amd64_Release_Image" ghcr.io/rook/rook-amd64:"$tag_amd64_Release_Image"
+    docker push ghcr.io/rook/rook-amd64:"$tag_amd64_Release_Image"
+
+    curl -LO https://github.com/estesp/manifest-tool/releases/download/v1.0.2/manifest-tool-linux-amd64
+    chmod +x manifest-tool-linux-amd64
+    sudo mv manifest-tool-linux-amd64 /usr/local/bin/manifest-tool
```
Comment (BlaineEXE) [inline tests/scripts/test-ci.sh:33]: If this is run locally, we don't want to move stuff into system locations. 

For GH actions, it might be fine, but I think it would still be better to leave the binary in the local dir and call it with `./manifest-tool`.

File: tests/scripts/test-ci.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+set -ex
+
+function push() {
+    arm64_Master_Image=$(docker images --format "{{.Repository}}:{{.Tag}}" | grep "arm64:latest")
```
Comment (BlaineEXE) [inline tests/scripts/test-ci.sh:5]: Grep will return multiple items without returning an error. We should consider how important it might be to check that we are only getting a single result. It might not be necessary since we are using `set -x`, but I would lean towards looking for the error so we don't accidentally push accidentally malformed image tags in unforeseen corner cases.

File: tests/scripts/test-ci.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+set -ex
```
Comment (BlaineEXE) [inline tests/scripts/test-ci.sh:2]:
```suggestion
set -xeuo pipefail
```

`set -euo pipefail` is the standard set of script options that most bash scripts start with. I like that you have `x` here for debugging also.

`pipefail` will make the script fail if a pipe step fails. E.g., if B fails here `A | B | C`, it counts as a script failure.

`u` will make sure any variables that are undefined also cause a failure, which is important to prevent things like `docker tag "$image" quay.io/tag:blah` from interpreting as `docker tag "" quay.io/tag:blah`

**Review by travisn — COMMENTED** [review]
Let's consider how to document where to find the images in different repos. I'm thinking wherever we use the rook image (operator.yaml, operator-openshift.yaml, etc), we can add a one-line comment with an example of other repos where they can be pulled.

**Comment by subhamkrai** [comment]
> Let's consider how to document where to find the images in different repos. I'm thinking wherever we use the rook image (operator.yaml, operator-openshift.yaml, etc), we can add a one-line comment with an example of other repos where they can be pulled.

I'll think about and work on the doc

**Comment by BlaineEXE** [comment]
There's also the `docker pulls` badge on the Readme. We could add quay and ghcr badges there also. Maybe it would be cleaner to separate the 3 container pull metric badges below the current ones. This could also be a future addition

**Comment by subhamkrai** [comment]
> There's also the `docker pulls` badge on the Readme. We could add quay and ghcr badges there also. Maybe it would be cleaner to separate the 3 container pull metric badges below the current ones. This could also be a future addition

I don't see img shiled has badges with pull information for quay and ghcr

File: .github/workflows/push-build.yaml [hunk 4]
```diff
@@ -18,13 +18,21 @@ permissions:
...
+    permissions:
+      packages: write
     if: github.repository == 'rook/rook'
     steps:
       - name: checkout
         uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
         with:
           fetch-depth: 0
 
+      - name: consider debugging
```
Comment (travisn) [inline .github/workflows/push-build.yaml:30]: If testing is done, we can remove this debugging step, right?

File: tests/scripts/test-ci.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+set -xeuo pipefail
+
+function push() {
+    arm64_Master_Image=$(docker images --format "{{.Repository}}:{{.Tag}}" | grep "arm64:latest" | head -n 1)
+    docker tag "$arm64_Master_Image" quay.io/rook/rook-arm64:master
+    docker push quay.io/rook/rook-arm64:master
```
Comment (travisn) [inline tests/scripts/test-ci.sh:7]: If we're in a release branch, we don't want to push the master image. We should only be pushing either to the master tag, release branch tag, or to the release version tag. Maybe the tag name could be passed as an arg from the makefile, so you don't have to compute it again.

File: tests/scripts/test-ci.sh [hunk 6]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
```
Comment (travisn) [inline tests/scripts/test-ci.sh:1]: Maybe rename the script `publish-build.sh`?

File: .github/workflows/push-build.yaml [hunk 7]
```diff
@@ -75,3 +97,7 @@ jobs:
           GITHUB_REF: $ {{ env.GITHUB_REF }}
         run: |
           tests/scripts/build-release.sh
+
+      - name: push to other OCI
+        run: |
+          tests/scripts/test-ci.sh push
```
Comment (travisn) [inline .github/workflows/push-build.yaml:103]: I know fully integrating with the makefile is more challenging, but can we at least call this script from the makefile? Then anytime `make publish` is called, we know all the resources are properly pushed. Our official builds are anyway only triggered by this workflow so it's not a big difference, but it still feels like it should be called from the makefile instead of from this workflow.

File: Documentation/Getting-Started/quickstart.md [hunk 8]
```diff
@@ -57,6 +57,9 @@ The first step is to deploy the Rook operator.
 !!! note
     These steps are for a standard production Rook deployment in Kubernetes. For Openshift, testing, or more options, see the [example configurations documentation](example-configurations.md).
 
+!!! note
+    Rook images are available in multiple container repositories. For dockerHub it is available under https://hub.docker.com/repository/docker/rook/ceph/tags org, for Quay it is available under https://quay.io/repository/rook/ceph and for ghcr images are available under https://github.com/orgs/rook/packages/container/package/ceph.
```
Comment (travisn) [inline Documentation/Getting-Started/quickstart.md:61]:
```suggestion
    Rook images are available in multiple container repositories. The examples by default reference [DockerHub](https://hub.docker.com/repository/docker/rook/ceph/tags). Images are also pushed to [Quay](https://quay.io/repository/rook/ceph?tab=tags) and [GHCR](https://github.com/orgs/rook/packages/container/package/ceph).
```

Comment (travisn) [reply Documentation/Getting-Started/quickstart.md:61]: The link for [GHCR](https://github.com/orgs/rook/packages/container/package/ceph) is not working, I guess it will be working after the first push from master?

File: tests/scripts/test-ci.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+function push() {
+    arm64_Master_Image=$(docker images --format "{{.Repository}}:{{.Tag}}" | grep "arm64:latest" | head -n 1)
+    docker tag "$arm64_Master_Image" quay.io/rook/rook-arm64:master
+    docker push quay.io/rook/rook-arm64:master
+    docker tag "$arm64_Master_Image" ghcr.io/rook/rook-arm64:master
+    docker push ghcr.io/rook/rook-arm64:master
+
+    arm64_Release_Image=$(docker images --format "{{.Repository}}:{{.Tag}}" | grep -v "arm64:latest" | grep -v "arm64:master" | grep arm64 | head -n 1)
+    tag_arm64_Release_Image="$(echo "$arm64_Release_Image" | cut -d ":" -f 2)"
```
Comment (travisn) [inline tests/scripts/test-ci.sh:12]: Could you factor out some sub-functions? There are a lot of repetitive commands that could be simplified.

**Review by travisn — CHANGES_REQUESTED** [review]

File: .github/workflows/push-build.yaml (hunk 4 above)
Comment (BlaineEXE) [reply .github/workflows/push-build.yaml:30]: This is a good point. Allowing tmate in the workflow that has access to image push secrets is a risk for our build and release pipeline.

File: tests/scripts/build-release.sh [hunk 10]
```diff
@@ -18,7 +18,7 @@ function publish() {
     $MAKE -C build/release build BRANCH_NAME=${BRANCH_NAME} TAG_WITH_SUFFIX=${TAG_WITH_SUFFIX} GIT_API_TOKEN=${GIT_API_TOKEN}
     git status &
     git diff &
-    $MAKE -C build/release publish BRANCH_NAME=${BRANCH_NAME} TAG_WITH_SUFFIX=${TAG_WITH_SUFFIX} AWS_ACCESS_KEY_ID=${AWS_USR} AWS_SECRET_ACCESS_KEY=${AWS_PSW} GIT_API_TOKEN=${GIT_API_TOKEN}
+    $MAKE -C build/release publish BRANCH_NAME=${BRANCH_NAME} TAG_WITH_SUFFIX=${TAG_WITH_SUFFIX} GIT_API_TOKEN=${GIT_API_TOKEN} AWS_ACCESS_KEY_ID=${AWS_USR} AWS_SECRET_ACCESS_KEY=${AWS_PSW}
```
Comment (BlaineEXE) [inline tests/scripts/build-release.sh:21]: So having `GIT_API_TOKEN` here with `set -x` above is going to leak these secrets I think. We need to be really careful with this. Please check to see if current GH logs are leaking this info.

If any creds have been leaked, we should rotate these ASAP.

File: tests/scripts/test-ci.sh [hunk 11]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
+set -xeuo pipefail
+
+function push() {
+    arm64_Master_Image=$(docker images --format "{{.Repository}}:{{.Tag}}" | grep "arm64:latest" | head -n 1)
```
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply tests/scripts/test-ci.sh:5]: I'm coming to this realization kind of late in this process, but one of the things I'm trying to puzzle out is why this script needs to list images in a repo in the first place. Usually, with image builds, the build system knows which image(s) it is building beforehand, then builds them. After that, tagging and combining images into manifests is easy because tags are pre-known.

This seems like a script that risks some amount of breakage by not assuming full responsibility for build, push, and manifest bundling. If the system that is pushing the images in the first place changes for some reason, then that could cascade into an issue here.

It's possible that this system is what is in place today, which is why it's being done like this today. Even if that's true, I think we should talk about whether it would be worth putting in effort to make our build+release system less fragile.

**Comment by onedr0p** [comment]
After the latest news of docker hub limiting unauthenticated pulls to 10/hour it would be great to see this completed.

**Comment by subhamkrai** [comment]
> After the latest news of docker hub limiting unauthenticated pulls to 10/hour it would be great to see this completed.

we'll try to get it done by EOW

File: tests/scripts/test-ci.sh (hunk 1 above)
Comment (subhamkrai) [reply tests/scripts/test-ci.sh]: done.

File: tests/scripts/build-release.sh (hunk 10 above)
Comment (subhamkrai) [reply tests/scripts/build-release.sh:21]: I think we are good here ```GIT_API_TOKEN=***```

Comment (BlaineEXE) [reply tests/scripts/build-release.sh:21]: Thanks for looking!

File: .github/workflows/push-build.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (subhamkrai) [reply .github/workflows/push-build.yaml]: thanks for suggestion since script is removed, this is not valid anymore.

File: tests/scripts/test-ci.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (subhamkrai) [reply tests/scripts/test-ci.sh]: thanks for the suggestion since the script is removed, this is not valid anymore.

File: tests/scripts/test-ci.sh (hunk 9 above)
Comment (subhamkrai) [reply tests/scripts/test-ci.sh]: thanks for suggestion since script is removed, this is not valid anymore.

File: Documentation/Getting-Started/quickstart.md (hunk 8 above)
Comment (subhamkrai) [reply Documentation/Getting-Started/quickstart.md]: Yess

File: .github/workflows/push-build.yaml (hunk 7 above)
Comment (subhamkrai) [reply .github/workflows/push-build.yaml]: done

File: tests/scripts/test-ci.sh (hunk 5 above)
Comment (subhamkrai) [reply tests/scripts/test-ci.sh]: thanks for suggestion since script is removed, this is not valid anymore.

**Comment by travisn** [comment]
As discussed, there is another CI issue going on, let's get the [publish action](https://github.com/rook/rook/actions/workflows/push-build.yaml) passing again before merging this...

**Comment by subhamkrai** [comment]
> > > There's also the `docker pulls` badge on the Readme. We could add quay and ghcr badges there also. Maybe it would be cleaner to separate the 3 container pull metric badges below the current ones. This could also be a future addition
> > 
> > 
> > I don't see img shiled has badges with pull information for quay and ghcr
> 
> @subhamkrai: Maybe we can make something up:
> 
> [badge](https://img.shields.io/badge/quay_pulls-500M-green.svg)

We can have this, but the data will not be real-time. And I'll like to add it in follow-up PR

**Review by BlaineEXE — APPROVED** [review]

**Comment by a1994sc** [comment]
I do not see the images under ghcr, you might need to make them public

**Comment by a1994sc** [comment]
Looks good!! So excited for this and the oci chart!!!
